app/Http/Controllers/Admin/CategoryController.php

```diff
@@ -3,18 +3,31 @@
 namespace App\Http\Controllers\Admin;
 
 use App\Http\Controllers\Controller;
+use App\Http\Requests\Admin\CategoryStoreRequest;
+use App\Repositories\Admin\Category\CategoryRepositoryInterface;
 use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
 
 class CategoryController extends Controller
 {
+    protected $categoryRepo;
+
+    public function __construct(CategoryRepositoryInterface $categoryRepo)
+    {
+        $this->categoryRepo = $categoryRepo;
+    }
```

Comment:
Trả lời câu hỏi: Inject Repository vào constructor như thế này có ưu điểm gì? Tại sao bn inject interface thay vì inject CategoryRepository hay gì đó?

Reply:
1. Inject Repository vào constructor như thế này có ưu điểm gì?
- Giúp tổ chức code clean hơn, nhìn dễ hiểu và trực quan hơn.
- Khi có thay đổi thì mình chỉ cần thay đổi trong contructor

2. Tại sao inject interface thay vì inject CategoryRepository?
- Vì đã binding Interface trong Provider mỗi một interface sẽ ứng với 1 repository tương ứng.
- Khi thay đổi database hoặc thay đổi hẳn base mới thì việc làm chúng ta chỉ cần tạo 1 interface mới rồi binding lại , đây được gọi là thiết kế nguyên lý phần mềm Inversion of Control (IOC)